File: Cricket Dahsboard/CD.cpp
#include <iostream>
#include <bits/stdc++.h>
using namespace std;

class Players
{

private:
    int Position, balls, runs, fours, sixes;

    int id;

public:
    Players(int Position, int balls, int runs, int fours, int sixes, int id)
    {
        this->id = id;
        this->Position = Position;
        this->balls = balls;
        this->runs = runs;
        this->fours = fours;
        this->sixes = sixes;
    }
    void setRuns(int Runs)
    {
        runs += Runs;
        if (Runs == 6)
            sixes++;
        if (Runs == 4)
            fours++;
    }

    void PlayedBalls()
    {
        balls++;
    }

    int getruns()
    {
        return runs;
    }

    int getPlayedballs()
    {
        return balls;
    }

    int getid()
    {
        return id;
    }

    int getsixes()
    {
        return sixes;
    }

    int getfours()
    {
        return fours;
    }

    int getposition()
    {
        return Position;
    }
};

class Cricket
{
    int players;
    int Overs;

public:
    Cricket(int players, int Overs)
    {
        this->players = players;
        this->Overs = Overs;
    }
    void PrintScoreCard(vector<Players> palyers, int score, int over, int balls, int p1, int p2, int wickets)
    {
        cout << "Name"
             << " "
             << "Runs"
             << " "
             << "Balls"
             << " "
             << "4's"
             << " "
             << "6's" << endl;
        for (int i = 0; i < palyers.size(); i++)
        {
            Players Item = palyers[i];
            cout << Item.getid();

            if (Item.getid() == p1 || Item.getid() == p2)
            {
                cout << "*"
                     << "  "; // On_strike
            }
            else
            {
                cout << "    ";
            }

            cout << "  " << Item.getruns() << "   " << Item.getPlayedballs() << "     " << Item.getfours() << "    " << Item.getsixes() << endl;
        }

        cout << "Score:" << score << "/" << wickets << endl;
        if (balls == 0)
        {
            cout << "Overs:" << over << endl;
        }
        else
        {
            cout << "Overs" << over << "." << balls << endl;
        }
    }

    pair<long, int> innings()
    {
        long Score = 0;
        queue<Players> team; // Declaring queue of Players Object
        vector<Players> scoreCard;

        // set the batting line up of players
        vector<int> lineup(players);
        cout << "Enter Batting line up:"
             << " ";
        for (int i = 0; i < players; i++)
        {
            cin >> lineup[i];
        }

        for (int i = 0; i < lineup.size(); i++)
        {

            team.push(Players(i, 0, 0, 0, 0, lineup[i]));
            scoreCard.push_back(Players(i, 0, 0, 0, 0, lineup[i]));
        }

        Players p1 = team.front(); // Creating object of two player
        team.pop();
        Players p2 = team.front();
        team.pop();

        Players On_Strike = p1; // creating object for striker end batsman for updaing the score.
        int Wickets = 0;
        for (int i = 1; i <= Overs; i++)
        {

            int balls = 6;

            while (balls > 0)
            {
                string run_on_this_ball;
                cout << "run_on_this_ball"
                     << " - ";
                cin >> run_on_this_ball;
                string runs = "0123456";

                if (runs.find(run_on_this_ball) != string::npos)
                {
                    long long val = stoi(run_on_this_ball);

                    On_Strike.setRuns(val);
                    On_Strike.PlayedBalls();
                    Score += val;

                    if (On_Strike.getid() == p1.getid())
                    {
                        p1 = On_Strike;

                        if (val % 2)
                            On_Strike = p2;
                    }
                    else
                    {
                        p2 = On_Strike;
                        if (val % 2)
                            On_Strike = p1;
                    }
                    balls--;
                }
                else if (run_on_this_ball == "W")
                {
                    Wickets++;
                    On_Strike.PlayedBalls();
                    scoreCard[On_Strike.getposition()] = On_Strike; // storing all the details for who got out in vector
                    if (On_Strike.getid() == p1.getid())
                    {
                        if (!team.empty())
                        {
                            p1 = team.front(); // new batsmam always takes strike.
                            team.pop();
                            On_Strike = p1;
                        }
                        // if team got all out
                        else
                        {

                            scoreCard[p2.getposition()] = On_Strike; // storing non striker batsam details;
                            PrintScoreCard(scoreCard, Score, i, (6 - balls) + 1, p2.getposition(), -1, Wickets);
                            return {Score, Wickets};
                        }
                    }
                    else
                    {
                        if (!team.empty())
                        {
                            p2 = team.front(); // new batsmam coming out on striker end
                            team.pop();
                            On_Strike = p2; // new batsman alwasy take strike
                        }
                        // if team got all out
                        else
                        {

                            scoreCard[p1.getposition()] = On_Strike; // storing non striker batsam details;
                            PrintScoreCard(scoreCard, Score, i, (6 - balls) + 1, p1.getposition(), -1, Wickets);
                            return {Score, Wickets};
                        }
                    }
                    balls--;
                }
                else if (run_on_this_ball == "WD")
                {
                    Score++;
                }
                cout << "Score:" << Score << "-" << Wickets << endl;
            }
            // Rotating the strike after an over
            if (On_Strike.getid() == p1.getid())
            {
                p1 = On_Strike;
                On_Strike = p2;
            }
            else
            {
                p2 = On_Strike;
                On_Strike = p1;
            }
            cout << "Over Completed:" << endl;
            scoreCard[p1.getposition()] = p1;
            scoreCard[p2.getposition()] = p2;
            PrintScoreCard(scoreCard, Score, i, 0, p1.getposition(), p2.getposition(), Wickets);
        }

        return {Score, Wickets};
    }
};

int main()
{

    int test_case;
    cout << "Test_Cases:-";
    cin >> test_case;

    while (test_case--)
    {

        int Overs, players;

        cout << "Overs"
             << ":";
        cin >> Overs;

        cout << "Players in each team"
             << ":";
        cin >> players;

        Cricket match(players, Overs);

        pair<long, int> team1 = match.innings();
        pair<long, int> team2 = match.innings();

        if (team1.first > team2.first)
        {
            cout << "team1 won the match";
        }
        else
        {
            cout << "team2 won the match";
        }
    }
};

File: Cricket Dahsboard/CD.py
class Players:
    def __init__(self, Position, balls, runs, fours, sixes, player_id):
        self.id = player_id
        self.Position = Position
        self.balls = balls
        self.runs = runs
        self.fours = fours
        self.sixes = sixes

    def set_runs(self, Runs):
        self.runs += Runs
        if Runs == 6:
            self.sixes += 1
        if Runs == 4:
            self.fours += 1

    def played_balls(self):
        self.balls += 1

    def get_runs(self):
        return self.runs

    def get_played_balls(self):
        return self.balls

    def get_id(self):
        return self.id

    def get_sixes(self):
        return self.sixes

    def get_fours(self):
        return self.fours

    def get_position(self):
        return self.Position


class Cricket:
    def __init__(self, players, overs):
        self.players = players
        self.overs = overs

    def print_score_card(self, players, score, over, balls, p1, p2, wickets):
        print("Name Runs Balls 4's 6's")
        for player in players:
            print(player.get_id(), '*' if player.get_id() == p1 or player.get_id() == p2 else ' ',
                  player.get_runs(), player.get_played_balls(), player.get_fours(), player.get_sixes())

        print(f"Score: {score}/{wickets}")
        if balls == 0:
            print(f"Overs: {over}")
        else:
            print(f"Overs: {over}.{balls}")

    def innings(self):
        Score = 0
        team = []
        score_card = []

        lineup = list(map(int, input("Enter Batting line up: ").split()))

        for i, player_id in enumerate(lineup):
            team.append(Players(i, 0, 0, 0, 0, player_id))
            score_card.append(Players(i, 0, 0, 0, 0, player_id))

        p1 = team.pop(0)
        p2 = team.pop(0)

        on_strike = p1
        wickets = 0

        for i in range(1, self.overs + 1):
            balls = 6

            while balls > 0:
                run_on_this_ball = input("Run on this ball: ")

                if run_on_this_ball in "0123456":
                    val = int(run_on_this_ball)
                    on_strike.set_runs(val)
                    on_strike.played_balls()
                    Score += val

                    if on_strike.get_id() == p1.get_id():
                        p1 = on_strike
                        on_strike = p2 if val % 2 else on_strike
                    else:
                        p2 = on_strike
                        on_strike = p1 if val % 2 else on_strike

                    balls -= 1

                elif run_on_this_ball == "W":
                    wickets += 1
                    on_strike.played_balls()
                    score_card[on_strike.get_position()] = on_strike

                    if on_strike.get_id() == p1.get_id():
                        if team:
                            p1 = team.pop(0)
                            on_strike = p1
                        else:
                            score_card[p2.get_position()] = on_strike
                            self.print_score_card(score_card, Score, i, (6 - balls) + 1, p2.get_position(), -1, wickets)
                            return Score, wickets

                    else:
                        if team:
                            p2 = team.pop(0)
                            on_strike = p2
                        else:
                            score_card[p1.get_position()] = on_strike
                            self.print_score_card(score_card, Score, i, (6 - balls) + 1, p1.get_position(), -1, wickets)
                            return Score, wickets

                    balls -= 1

                elif run_on_this_ball == "WD":
                    Score += 1

                print(f"Score: {Score} - {wickets}")

            if on_strike.get_id() == p1.get_id():
                p1 = on_strike
                on_strike = p2
            else:
                p2 = on_strike
                on_strike = p1

            score_card[p1.get_position()] = p1
            score_card[p2.get_position()] = p2
            self.print_score_card(score_card, Score, i, 0, p1.get_position(), p2.get_position(), wickets)

        return Score, wickets


if __name__ == "__main__":
    test_case = int(input("Test Cases: "))

    while test_case > 0:
        overs = int(input("Overs: "))
        players = int(input("Players in each team: "))

        match = Cricket(players, overs)
        team1_score, team1_wickets = match.innings()
        team2_score, team2_wickets = match.innings()

        if team1_score > team2_score:
            print("Team 1 won the match")
        else:
            print("Team 2 won the match")

        test_case -= 1
